### PR TITLE
Switch to grafana-openapi-client-go

### PR DESCRIPTION
**Pull request by mbarrien**
Switches datasource, dashboard, and folder resources to using the grafana-openapi-client-go instead of accessing the REST API's directly. This should supersede #260 (and possibly #257 since it adds a field to provider for the OpenAPI client) 

There are some small issues:
* The OpenAPI models JSON rendering has a lot of `omitempty` annotations, which results in many fields no longer being present in the next pull after you change to this code (especially for datasources).
* Some tests not passing `make test`:
  * Some bugs in the OpenAPI spec
    * getDatasourceByName does not define a 404 response, so in Golang code had to fall back to runtime.APIError instead of an error specific to that function call.
    * postDashboard the `id` field is expected to be a string, but Grafana is return a number
  * Getting 409 and 412 errors on some tests which aren't expecting them.

Note that some of the functionality lost in #260 due to missing fields has been restored since grafana-openapi-client-go contains them all.

If you have previously run grizzly pull, then run grizzly pull again in the new code, you will get diffs like this on all previously pulled resources (anonymized from actual examples):

Datasources:
```
--- a/grafana/grizzly/datasources/datasource-Grafana Cloud k6.yaml
+++ b/grafana/grizzly/datasources/datasource-Grafana Cloud k6.yaml
@@ -4,11 +4,7 @@ metadata:
     name: Grafana Cloud k6
 spec:
     access: proxy
-    basicAuth: false
-    basicAuthUser: ""
-    database: ""
     id: 33
-    isDefault: false
     jsonData:
         backendUrl: https://k6.io
         environment: production
@@ -18,11 +14,8 @@ spec:
     name: Grafana Cloud k6
     orgId: 1
     readOnly: true
-    secureJsonFields: {}
     type: k6-datasource
     typeLogoUrl: /public/plugins/k6-app/img/logo.svg
     uid: Grafana Cloud k6
     url: proxy
-    user: ""
     version: 1
-    withCredentials: false
```

Dashboards:
```
--- a/grizzly/dashboards/example-folder/dashboard-bazqux.yaml
+++ b/grizzly/dashboards/example-folder/dashboard-bazqux.yaml
@@ -23,6 +23,7 @@ spec:
     editable: true
     fiscalYearStartMonth: 0
     graphTooltip: 0
+    id: 180
     links: []
     liveNow: false
     panels:
@@ -133,4 +134,5 @@ spec:
     timezone: ""
     title: Sample Dashboard
     uid: bazqux
+    version: 2
     weekStart: ""
```

Folders:
```
--- a/grafana/grizzly/folders/folder-foobar.yaml
+++ b/grafana/grizzly/folders/folder-foobar.yaml
@@ -3,17 +3,16 @@ kind: DashboardFolder
 metadata:
     name: da4SikgVk
 spec:
-    canAdmin: false
     canDelete: true
     canEdit: true
     canSave: true
-    created: "2022-07-14T12:11:57Z"
+    created: "2022-07-14T12:11:57.000Z"
     createdBy: Anonymous
-    hasAcl: false
     id: 3
+    parents: null
     title: GrafanaCloud
     uid: foobar
-    updated: "2022-07-14T12:11:57Z"
+    updated: "2022-07-14T12:11:57.000Z"
     updatedBy: Anonymous
     url: /dashboards/f/foobar/example
     version: 1
```

### CODE REVIEW

**Comment by joanlopez** [comment]
Hi there @mbarrien,

Thanks for your effort on helping to migrate to the new client. I think it'd be awesome to get this pull request ready to be merged. So, beyond an eventual code review, now I'm going over all the failing tests (the ones you mentioned), case by case, and trying to fix them (see for instance 👉🏻  https://github.com/grafana/grafana/pull/76749).

Once ready, I'll create a PR against your branch, so hopefully you'll branch will be ready to be merged.

Once more, thanks for your contributions, I think we're getting closer to finally have the "good" client in place! 🙌🏻

File: pkg/grafana/dashboards.go [hunk 1]
```diff
@@ -1,296 +1,115 @@
...
-	}
+// Moved from utils.go
+const generalFolderId = 0
+const generalFolderUID = "general"
 
-	req, err := http.NewRequest("GET", grafanaURL, nil)
-	if err != nil {
-		return nil, err
-	}
+// Losing id, typeLogoUrl, version, withCredentials
```
Comment (malcolmholmes) [inline pkg/grafana/dashboards.go:20]: Is this line needed? Or suggests something still needs doing?

File: pkg/grafana/dashboards.go [hunk 2]
```diff
@@ -1,296 +1,115 @@
...
-	switch {
-	case resp.StatusCode == http.StatusNotFound:
-		return nil, grizzly.ErrNotFound
-	case resp.StatusCode >= 400:
-		return nil, errors.New(resp.Status)
-	}
-
-	data, err := io.ReadAll(resp.Body)
+	// TODO: Turn spec into a real models.DashboardFullWithMeta object
+	spec, err := structToMap(dashboard.Dashboard)
```
Comment (malcolmholmes) [inline pkg/grafana/dashboards.go:36]: We have github.com/mitchellh/mapstructure that can do this too.

File: pkg/grafana/datasources.go [hunk 3]
```diff
@@ -1,205 +1,110 @@
...
-		return nil, errors.New(resp.Status)
-	}
-
-	data, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
-	}
-	return data, nil
-}
+// Losing a bunch of omitempty fields
```
Comment (malcolmholmes) [inline pkg/grafana/datasources.go:16]: unnecessary comment?

File: pkg/grafana/utils.go [hunk 4]
```diff
@@ -1,29 +1,44 @@
...
+		return nil, err
+	}
+
+	var result map[string]interface{}
+	if err := json.Unmarshal(jsonData, &result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
```
Comment (malcolmholmes) [inline pkg/grafana/utils.go:44]: as said above, can we use github.com/mitchellh/mapstructure for this?

**Review by malcolmholmes — COMMENTED** [review]
Thanks for this, much appreciated. A few small comments/questions, but I think we can approve merge this quite soon.

File: pkg/grafana/dashboards.go (hunk 2 above)
Comment (joanlopez) [reply pkg/grafana/dashboards.go:36]: I'm not sure if it's as easy as that, see 👉🏻  https://github.com/grafana/grizzly/pull/257#discussion_r1352778493

File: pkg/grafana/utils.go (hunk 4 above)
Comment (joanlopez) [reply pkg/grafana/utils.go:44]: Ditto

File: pkg/grafana/dashboards.go (hunk 2 above)
Comment (malcolmholmes) [reply pkg/grafana/dashboards.go:36]: thanks for these pointers, can go with the code in the PR then.

**Review by malcolmholmes — APPROVED** [review]

**Pull request merged**